tensorflow/core/kernels/mkl_qmatmul_op.cc

```diff
@@ -0,0 +1,875 @@
+//           filter min-max values.
+// Other than that, this op does not support other input combination yet.
+// When input is quantized to uint8 via MIN_FIRST, bias needs compensation.
+// The detailed algorithm is illustrated as below:
+//
+// A𝑓32 is the original fp32 activation 2D tensor.
+// Min(A𝑓32) is the minimum scalar value of A𝑓32.
+// Max(A𝑓32) is the maximum scalar value of A𝑓32.
+// MaxAbs(A𝑓32) is the maximum absolute scalar value of A𝑓32.
+// Qa is the quantization scale for activation.
+// Au8 is the quantized unsigned int8 activation tensor.
+// With SCALE quantization (used for non-negative A𝑓32), Qa and Au8 can be
+// calculated as below:
+//    Qa = 255.0 / Max(A𝑓32)
+//    Au8 = round(Qa * A𝑓32).
+// With MIN_FIRST quantization, Q'a and A'u8 can be calculated as below:
+//    Q'a = 255.0 / (Max(A𝑓32) – Min(A𝑓32))
+//    A'u8 = round(Q'a * (A𝑓32 – Min(A𝑓32) * ones(A𝑓32))),
+// where, ones(.) is a tensor of all 1s with the same shape of its argument and
+// round(.) rounds a number to its nearest integer.
+//
+// W𝑓32 is the original fp32 2D weight tensor.
+// MaxAbs(W𝑓32) is the maximum absolute scalar value of W𝑓32.
+// Qw is the quantization scale of weight.
+// Ws8 is the quantized signed int8 weight tensor.
+// Qw and Ws8 can be calculated as below:
+//    Qw = 127.0 / MaxAbs(W𝑓32)
+//    Ws8 = round(Qw * W𝑓32).
+//
+// B𝑓32 is the original fp32 1D bias tensor matching the innermost dim of W𝑓32.
+// With SCALE quantization of activation, the scaled bias, Bs32, is calculated
+// as below:
+//      Bs32 = Qa * Qw * B𝑓32.
+// With MIN_FIRST quantization of activation, the scaled bias tensor with
+// compensation, B's32, is calculated as below:
+//      B's32 = Q'a * Qw * B𝑓32 + Q'a * Qw * Min(A𝑓32) * 1 * W𝑓32
+// where, 1 denotes a row vector matching the outermost dim of W𝑓32.
+//
+// The QuantizedMatMulWithBias op calculates 32bit integer output as below:
+//  - with SCALE activation quantizaiton:
+//    Xs32 = Au8 * Ws8 + 1' * Bs32
+//         = Qa * Qw * A𝑓32 * W𝑓32  + Qa * Qw * 1' * B𝑓32
+//         = Qa * Qw * (A𝑓32 * W𝑓32 + 1' * B𝑓32) = Qa * Qw * X𝑓32,
+//    where, 1' denotes a column vector matching the outermost dim of A𝑓32 and
+//    X𝑓32 represents the output of original fp32 MatMul with BiasAdd fusion.
+//
+//  - with MIN_FIRST activation quantization:
+//    Xs32 = A'u8 * Ws8 + 1' * B's32
+//         = Q'a * (A𝑓32 - Min(A𝑓32) * ones(A𝑓32)) * Qw * W𝑓32 +
+//           Q'a * Qw * 1' * B𝑓32 + Q'a * Qw * Min(A𝑓32) * 1' * 1 * W𝑓32
+//         = Q'a * Qw * (A𝑓32 * W𝑓32 + 1' * B𝑓32)
+//         = Q'a * Qw * X𝑓32.
+//    Note that 1' * 1 = ones(A𝑓32).
+//
+// The QuantizedMatMulWithBiasAndRelu op does the same calucation as above
+// except adding relu function for the 32bit integer output.
+//
+// The QuantizedMatMulWithBiasAndReluAndRequantize op do one more requantize
+// calculation based on above. The requantize scale Qr is calculated from
+// offline calibration.
+//    Qr = 255 / MaxAbs(X𝑓32)
```

Comment:
@penpornk Addressed in the latest commit. Note that since original fp32 graph has relu as the end node for fusion candidate, Xf32 is guaranteed to be non-negative. Hence replaced with Qr = 255 / Max(Xf32).

Reply:
Thank you for the explanation. Could you please add in the comment that Xf32 is guaranteed to be positive because of Relu? I'll add a comment in the current code since this comment is outdated.